### PR TITLE
Exposure the race condition on the pod preemption

### PR DESCRIPTION
**Pull request by chendave**
- Fix a potential test flake.
- Add new testcase to exposure the race condition on pod preemption.
- Take the chance to fix a typo.

Signed-off-by: Dave Chen <dave.chen@arm.com>

**What type of PR is this?**

Add one of the following kinds:
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Ref https://github.com/kubernetes/kubernetes/issues/93505

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by chendave** [comment]
/cc @Huang-Wei  @ahg-g  @alculquicondor  @soulxu 

I will propose a fix later, the preemption could eventually got convergent if we skip the move rquest when the `unscheduleQ` is empty.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 1]
```diff
@@ -363,12 +363,69 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
 	q.Add(&medPriorityPod)
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&unschedulablePod), q.SchedulingCycle())
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
+	// sleep 1 second so that pods is guaranteed to be added into unschedulableQ first and then
+	// the move request is happened later, otherwise there is potential a test flake here, as the
+	// move request has a chance to be finished before the pods get into the unschedulableQ.
+	// In that case, the pods will be moved to backoffQ even `QueuedPodInfo.Attempts` is zero.
+	time.Sleep(1000 * time.Millisecond)
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:370]: Do not do this. Instead, create the queue with a fake clock: `q := NewPriorityQueue(newDefaultQueueSort(), WithClock(clock.NewFakeClock(time.Now())))`

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 2]
```diff
@@ -363,12 +363,69 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
...
+	}
+}
+
+// TestPriorityQueue_RaceConditionOnScheduling mimic a race condition could be found with preemption.
+// In this particular case, the high priority pod is stuck in the backoffQ and if new pod is coming,
+// regardless it's priority, the new pod will be scheduled first.
+// The preemption could happen again but there is a chance the pod will never got scheduled as long as
+// there is new pod coming for scheduling since the high priority pod could be moved from backoffQ
+// to activeQ only when the time for backoff expiration is up.
+// see https://github.com/kubernetes/kubernetes/issues/93505 for the details.
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:386]: I'm not sure how is this related. Could you document your findings in the issue?

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 3]
```diff
@@ -363,12 +363,69 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
...
+	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
 	if q.activeQ.Len() != 1 {
 		t.Error("Expected 1 item to be in activeQ")
 	}
-	if q.podBackoffQ.Len() != 2 {
-		t.Error("Expected 2 items to be in podBackoffQ")
+	if q.podBackoffQ.Len() != 1 {
+		t.Error("Expected 1 items to be in podBackoffQ")
+	}
+	// the lowPriority pod is popped and got scheduled while the highPriorityPod is stuck in the
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:408]: This is by design to avoid starvation from high priority pods. Let's discuss in the original issue. However, we intend to reduce the problems of this with #94009

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:370]: Done

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 3 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:408]: problem here is that the `highPriorityPod` should not be in `backoffQ` in the first place, it is moved to be `backoffQ` just because there was a move request happened first.
`lowPriority` pod popped here should be fine, but there should a way for the preemption to get convergent,  pls see the code following, 
```golang
	// another pod is added to activeQ.
	q.Add(&unschedulablePod)
	if q.activeQ.Len() != 1 {
		t.Error("Expected 1 item to be in activeQ")
	}
```
it's possible a new pod is added to Queue while the `highPriorityPod` is still backoff-ing, after the backoff is up, the `highPriorityPod` needs to preempt the `lowPriority` pod again.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 2 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:386]: updated in that issue, pls kindly check it at your convenient time.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 4]
```diff
@@ -359,16 +359,76 @@ func TestPriorityQueue_Delete(t *testing.T) {
...
+	c := clock.NewFakeClock(time.Now())
+	q := NewPriorityQueue(newDefaultQueueSort(), WithClock(c))
 	q.Add(&medPriorityPod)
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&unschedulablePod), q.SchedulingCycle())
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
+	// Move clock by 1 second so that pods is guaranteed to be added into unschedulableQ first and
+	// then the move request is happened later, otherwise there is potential a test flake here, as
+	// the move request has a chance to be finished before the pods get into the unschedulableQ.
+	// In that case, the pods will be moved to backoffQ even `QueuedPodInfo.Attempts` is zero.
+	c.Step(time.Second)
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:371]: Could you rather leave the original test case without the Step?

Or we could have 2 tests.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:411]: I prefer this to be split in a different PR so we can already merge the tests using the deterministic clock.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 5]
```diff
@@ -359,16 +359,76 @@ func TestPriorityQueue_Delete(t *testing.T) {
...
+	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
+	if q.activeQ.Len() != 1 {
+		t.Error("Expected 1 item to be in activeQ")
+	}
+	if q.podBackoffQ.Len() != 1 {
+		t.Error("Expected 1 items to be in podBackoffQ")
+	}
+	// the lowPriority pod is popped and got scheduled while the highPriorityPod is stuck in the
+	// backoffQ.
+	q.Pop()
```
Comment (soulxu) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:413]: I still stuck at the problem of the low priority pod shouldn't be gotten a chance to schedule successful, since when the low priority pod getting scheduling, the resources already token by high priority pod.

Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:413]: please thinking about the case that high priority pod is still in the `unscheduleQ` or `backoffQ`, low priority pod is the only one in the `activeQ`.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 4 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:371]: Done, it's here: https://github.com/kubernetes/kubernetes/pull/94413
